### DISCUSSION TITLE
New features and breaking changes in 0.12.x

### DISCUSSION THREAD

**Discussion by lefou** [com-lihaoyi/mill#3349] [News]
Below you find a new feature or a breakage per main-comment. Discussion should happen in the matching sub-comments.

The development of Mill `0.12` starts with the tag `0.12.0-M0` in the `main` branch. It is planned to keep binary compatibility to Mill `0.11`, so that you can keep using plugins built for Mill `0.11`. Therefore we also keep the binary platform to `0.11`, which means, thirdparty plugins will keep using the `_mill0.11` suffix in their artifact name.

Maintenance of Mill `0.11` will happen for a rather short time in the `0.11.x` branch, as we hope that it's easy for most users to just switch over to `0.12`. 

Please open a __new__ issue or a discussion, if you face major road blocks when switching from `0.11` to `0.12`!

**Comment by lefou** [comment]
Mill `0.12` will requires Java 11.

* https://github.com/com-lihaoyi/mill/pull/3282

**Comment by lefou** [comment]
Mill `0.12` supports submodules defined in ~~`module.sc`~~ `package.mill` files in sub-directories.

* https://github.com/com-lihaoyi/mill/pull/3213

**Comment by lefou** [comment]
Mill `0.12` now enables parallel builds by default.

* https://github.com/com-lihaoyi/mill/pull/3265

**Comment by lihaoyi** [comment]
Mill `0.12` runs test suites with the subprocess working directory set to empty sandbox folders by default, rather than in the project root

* https://github.com/com-lihaoyi/mill/pull/3347

**Comment by lihaoyi** [comment]
We'll also be moving the Mill process `os.pwd` to a sandbox folder by default, to encourage proper use of `T.source`s, `T.dest` and `PathRef`s rather than directly reading files off of disk

* https://github.com/com-lihaoyi/mill/pull/3367

**Comment by lefou** [comment]
Dropped Support for Scala Native `0.4`

* as part of https://github.com/com-lihaoyi/mill/pull/3398

**Comment by lefou** [comment]
The name of the `BaseModule` is no longer `build`, but `build.package`. 

As a result, when using `mill.resolve.Resolve.resolve` API, you may need to change your code.

A possible code change might look like this:

```diff
 mill.resolve.Resolve.Tasks.resolve(
-    build,
+    build.`package`,
     taskNames,
     SelectMode.Separated
   )
```

**Reply by lihaoyi** [reply]
This should be no longer necessary once https://github.com/com-lihaoyi/mill/pull/3426 lands

**Reply by lefou** [reply]
Nice to see that we don't introduce extra use of package objects. While experimenting, I think I witnessed various strange incremental compiling issues caused by package objects.

**Reply by lihaoyi** [reply]
Yeah I hit a bunch of issues. I assume some are fixed in Scala3, and package objects do seem to fit the domain nicely, but some issues remain (e.g. the `build.package` thing) so maybe best to go eith normal objects for now

**Comment by lolgab** [comment]
`forkEnv` is now an empty `Map` #3442

**Comment by lefou** [comment]
Mill commands now require arguments to be passed named via `./mill mycommand --key value`, rather than allowing just `./mill mycommand value.` 

You can pass in `--allow-positional-command-args` to fall back to the old behavior, or use `@mainargs.arg(positional = true)` on individual parameters

* https://github.com/com-lihaoyi/mill/pull/3431

**Comment by lihaoyi** [comment]
The type parameter for `Discover[T]` was removed in https://github.com/com-lihaoyi/mill/commit/40b19204dd9a40448bc284ce5394748c80a40c75, leaving just `Discover`. The type parameter was unused, and removing it is binary compatible although not source compatible. For migration purposes, you should be able to convert any `Discover[T]` types to just `Discover` and things should continue working as before

**Comment by gamlerhart** [comment]
I maybe missed it somewhere, but a breaking change also?  
The `RootModule` must now be called `package`.

**Reply by lefou** [reply]
Correct. And since `package` is a reserved keyword, you need to mask it in Scala.

```scala
object `pakcage` extends RootModule {
}
```

**Reply by nafg** [reply]
Couldn't that just be

```scala
package object whatever extends RootModule
```
?

**Reply by lefou** [reply]
@nafg No. We evaluated that, but it has issues. Also, package objects will go in Scala 3.

**Comment by lefou** [comment]
`JavaModule.runIvyDeps` are now transitive.

* https://github.com/com-lihaoyi/mill/issues/3761
* https://github.com/com-lihaoyi/mill/pull/3774

**Comment by Tang-Haojin** [comment]
I have a file `submodule/common.sc`, and in previous version, I can use `import $file.submodule.common`. Now I get

```
[error] object submodule is not a member of package build_
```

How should I modify my `build.mill`?

**Reply by lihaoyi** [reply]
@Tang-Haojin you may be missing an `import $packages._` in your build.mill, or a `package.mill` file inside the `submodule/` folder

**Reply by Tang-Haojin** [reply]
Thanks! Here is another question: there are many `build.sc` in our submodules and previously in mill 0.11.x, we just ignore them. But now, mill reports something like `generateScriptSources Mill build.sc files can only be in the project root`. Is there an option to just ignore them rather than throwing an error?

Take https://github.com/chipsalliance/rocket-chip as an example. There are many submodules under `dependencies` directory, and they contains their own `build.sc`, but we do not actually use them in rocket-chip.

**Reply by Tang-Haojin** [reply]
Or are there some better ways to handle such circumstances?

**Reply by lefou** [reply]
Ignoring them should be safe. There is no way to collide, unless we support running from sub-modules. But this is an unwritten story, so it feels premature to prevent such situations. (The only situation could be that a sub-directory contains a `package.sc` and a `build.sc` with enabled sub-module scanning. Than both projects might share deeper sub-modules, but we need to discuss, if this is an problem or not. It could be a feature, too.)

**Reply by KireinaHoro** [reply]
Is there a way to downgrade this error into a warning?  I'm also encountering the same `generateScriptSources Mill build.sc files can only be in the project root` error on upgrading from 0.11.x.  The difference is that I'll be importing these build.sc files inside my build, to use them as foreign modules.

**Reply by lefou** [reply]
Not yet, but if we agree, a `build.{sc,mill}` in a sub-directory isn't an issue, we could change it to a warning. @lihaoyi WDYT?

**Reply by lihaoyi** [reply]
Works for me

**Reply by Readon** [reply]
Thanks. Would you put it in the next release?

**Reply by lefou** [reply]
It depends on how fast someone can prepare a PR. I think the next release is rather ripe.

**Reply by lefou** [reply]
I opened an issue

* https://github.com/com-lihaoyi/mill/issues/3921

**Reply by lefou** [reply]
Next Mill `0.12.3` will not longer fail in such situations.

* https://github.com/com-lihaoyi/mill/pull/3954

**Reply by Emin017** [reply]
Although mill now doesn't throw exceptions when it found `build.{sc,mill,mill.scala}` in subdir, I've found that mill now doesn't support importing `build.sc` from subdir, which would prevent projects like [rocket-chip](https://github.com/chipsalliance/rocket-chip) and [chisel-nix](https://github.com/chipsalliance/chisel-nix) from completing the build.
I think maybe the problem is here:
https://github.com/com-lihaoyi/mill/blob/a8fea1e947b1ad3b50f46d51fa01a6ea4c83737b/runner/src/mill/runner/CodeGen.scala#L33

**Reply by lefou** [reply]
@Emin017 What's wrong with this line? Do you have a reproducer?

**Reply by Emin017** [reply]
@lefou I'm trying to import `xxx.sc` from subdir (dependencies/xxx/xxx.sc) like this:
```scala
import mill._
import mill.scalalib._
import mill.scalalib.publish._
import coursier.maven.MavenRepository
import $file.dependencies.hardfloat.{common => hardfloatCommon}
import $file.dependencies.cde.{common => cdeCommon}
import $file.dependencies.diplomacy.{common => diplomacyCommon}
import $file.dependencies.chisel.{build => chiselBuild} // It seems only this import fail
import $file.{common => commonModule}
```
But I got the following error:
```shell
mill emulator[freechips.rocketchip.system.TestHarness,freechips.rocketchip.system.DefaultConfig].mfccompiler.compile
[build.sc-64/68] compile
[build.sc-64] [info] compiling 6 Scala sources to /home/nixos/Workbench/rocket-chip/out/mill-build/compile.dest/classes ...
[build.sc-64] [error] /home/nixos/Workbench/rocket-chip/build.sc:8:8: object build is not a member of package build_.dependencies.chisel
[build.sc-64] [error] import build_.dependencies.chisel.{build => chiselBuild}
[build.sc-64] [error]        ^
[build.sc-64] [error] /home/nixos/Workbench/rocket-chip/build.sc:32:11: not found: value chiselBuild
[build.sc-64] [error]   extends chiselBuild.Chisel {
[build.sc-64] [error]           ^
```
Here is my full [build.sc](https://github.com/user-attachments/files/18243748/build.sc.txt) for [rocket-chip](https://github.com/chipsalliance/rocket-chip) project (I'm trying to migrate it from 0.11.x to 0.12.x)

I have read some of the Mill source code, and I suspect the issue may be that Mill is skipping the search for the `build.sc` file in the subdirectories. However, I am still quite unfamiliar with the details of the Mill source code, so this is just my best guess. If I have made any incorrect assumptions, please accept my apologies.

**Reply by Emin017** [reply]
> @Emin017 What's wrong with this line? Do you have a reproducer?

I wrote the minimal reproducible demo here: https://github.com/Emin017/mill-multi-file-bug

**Reply by lihaoyi** [reply]
Mill skips `build.sc` files in sub-directories because it assumes that the sub-directory belongs to a different Mill project.

Is renaming the `build.sc` file something else an option for you guys? Any other name would do, as long as it's not `build.sc` or `build.mill`

**Reply by Emin017** [reply]
> Mill skips `build.sc` files in sub-directories because it assumes that the sub-directory belongs to a different Mill project.
> 
> Is renaming the `build.sc` file something else an option for you guys? Any other name would do, as long as it's not `build.sc` or `build.mill`

Thank you for the response! I think we should look into refactoring our build system then (Or we will have a discussion on it). And I'm very sorry for the inconvenience.

**Comment by Quafadas** [comment]
I hit a case where tests classes failed on 0.12.0, but worked on 0.11.12. Worked in CI, but not locally, which was odd.

In my build, I was adding this flag to forkArgs.

```
val vecIncubatorFlag = Seq("""--add-modules=jdk.incubator.vector""")
```

but hitting this;

```
1 tasks failed
vecxt.jvm.test.discoveredTestClasses java.lang.IllegalAccessError: superclass access check failed: class jdk.incubator.vector.Vector (in unnamed module @0x24e544f9) cannot access class jdk.internal.vm.vector.VectorSupport$Vector (in module java.base) because module java.base does not export jdk.internal.vm.vector to unnamed module @0x24e544f9
```

The workaround seems to be adding the jvm flag to mill opts, but this might be expected behaviour. Discord discussion;
https://discord.com/channels/632150470000902164/1295786923410984981/1299305997898354749

```
The error looks like it is coming from the Mill JVM before it forks off the test subprocess
from the discoveredTestClasses task
discoveredTestClasses runs test discovery in-process before forking off the subprocesses
not sure why it's causing issues, but you can probably solve it by adding --add-modules=jdk.incubator.vector to .mill-jvm-opts
```

**Comment by Quafadas** [comment]
Thank you!

**Comment by nafg** [comment]
Are there migration instructions somewhere?

**Comment by nafg** [comment]
Seems like it now reads all `.sc` files in the base directory?
